numpy/core/src/common/simd/intdiv.h

```diff
@@ -0,0 +1,485 @@
+        L2  = (npy_uint16)(1 << L);               // 2^L, overflow to 0 if L = 16
+        m   = ((L2 - d) << 16) / d + 1;           // multiplier
+        sh1 = 1;  sh2 = L - 1;                    // shift counts
+    }
+    npyv_u16x3 divisor;
+    divisor.val[0] = npyv_setall_u16(m); // multiplier
+#ifdef NPY_HAVE_SSE2 // SEE/AVX2/AVX512
+    divisor.val[1] = npyv_set_u16(sh1);  // shift count 1
+    divisor.val[2] = npyv_set_u16(sh2);  // shift count 2
+#elif defined(NPY_HAVE_VSX2)
+    divisor.val[1] = npyv_setall_u16(sh1);  // shift count 1
+    divisor.val[2] = npyv_setall_u16(sh2);  // shift count 2
+#elif defined(NPY_HAVE_NEON)
+    divisor.val[1] = npyv_reinterpret_u16_s16(npyv_setall_s16(-sh1));  // shift count 1
+    divisor.val[2] = npyv_reinterpret_u16_s16(npyv_setall_s16(-sh2));  // shift count 2
+#else
+    #error "New arch?"
+#endif
+    return divisor;
+}
+
+// encapsulate parameters for fast division on vector of 16-bit signed integers
+NPY_FINLINE npyv_s16x3 npyv_divisor_s16(npy_int16 d)
+{
+    int d1 = abs(d);
+    int sh, m;
+    if (d1 > 1) {
+        // shift count = ceil(log2(d1))-1 = (bitscan_revnz_u32(d1-1)+1)-1
+        sh = npyv__bitscan_revnz_u32(d1-1);
+        // calculate multiplier
+        m = (1U << (16 + sh)) / d1 + 1;
+    }
+    else {
+        m = 1; // for d1 = 1
+        sh = 0;
+        if (d == 0) {
+            // provoke error here if d = 0
+            m /= ((npy_int16 volatile *)&d)[0];
+        }
+    }
+    npyv_s16x3 divisor;
+    divisor.val[0] = npyv_setall_s16(m);              // broadcast multiplier
+    divisor.val[2] = npyv_setall_s16(d < 0 ? -1 : 0); // sign of divisor
+#ifdef NPY_HAVE_SSE2 // SEE/AVX2/AVX512
+    divisor.val[1] = npyv_set_s16(sh);                // shift count
+#elif defined(NPY_HAVE_VSX2)
+    divisor.val[1] = npyv_setall_s16(sh);             // shift count
+#elif defined(NPY_HAVE_NEON)
+    divisor.val[1] = npyv_setall_s16(-sh);            // shift count
+#else
+    #error "New arch?"
+#endif
+    return divisor;
+}
+// encapsulate parameters for fast division on vector of 32-bit unsigned integers
+NPY_FINLINE npyv_u32x3 npyv_divisor_u32(npy_uint32 d)
+{
+    npy_uint32 L, L2, sh1, sh2, m;
+    switch (d) {
+    case 0:
+        // provoke error for d = 0
+        m = sh1 = sh2 = 1 / ((npy_uint32 volatile *)&d)[0];
+        break;
+    case 1:
+        m = 1; sh1 = sh2 = 0;
+        break;
+    case 2:
+        m = 1; sh1 = 1; sh2 = 0;
+        break;
+    // general case for d > 2
+    default:
+        L   = npyv__bitscan_revnz_u32(d - 1) + 1;     // ceil(log2(d))
+        L2  = (npy_uint32)(1ULL << L);                // 2^L, overflow to 0 if L = 32
+        m   = ((npy_uint64)(L2 - d) << 32) / d + 1;   // multiplier
+        sh1 = 1;  sh2 = L - 1;                        // shift counts
+    }
+    npyv_u32x3 divisor;
+    divisor.val[0] = npyv_setall_u32(m); // multiplier
+#ifdef NPY_HAVE_SSE2 // SEE/AVX2/AVX512
```

Comment:
The shift count operation in npyv_divisor_u8/u16/u32 is similiar, can you make a macro functon?

Reply:
>  can you make a macro function?

I prefer to leave some room for flexibility for the upcoming SIMD extensions